design-documents/webhooks.md

```diff
@@ -0,0 +1,160 @@
+* Webhook framework notifies all subscribers with customer data
+
+:warning: Webhooks are supposed to push data updates only. Initial data should be extracted from Magento using standard APIs like GrapQL and REST. :warning:
+
+#### Webhook declaration
+New webhook could be defined by the following XML declaration in module context:
+
+```xml
+<!-- catalog_admin/etc/webhooks.xml -->
+<webhooks>
+    <webhook name="customer_updated" payload="\Magento\Customer\Api\Data\CustomerQueryInterface" />
+</webhooks>
+```
+This declaration will be used as a unique identifier of payload and unique identifier for the subscription.
+Also, having defined payload allows to generate webhooks documentation dynamically.
+
+#### Event
+Webhook listeners may listen to standard Magento events in order to collect changed entities:
+```xml
+<!-- Magento/Customer/etc/adminhtml/events.xml -->
+
+<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Event/etc/events.xsd">
+    <event name="customer_register_after">
+        <observer name="customer_register_webhook" instance="Magento\Customer\Oberservers\RegisterWebhook" />
+    </event>
+</config>
+```
+
+#### Webhook emission
+The central place of webhooks processing is webhook listeners. The responsibility of webhook listener is to identify changed entity, extract related data form query API and push the data into webhooks pipeline.
+```php
+// Magento/Customer/Observers/RegisterWebhook.php
+use \Magento\Framework\Webhook\Dispatcher;
+
+class RegisterWebhook
+{
+    public function execute(\Magento\Framework\Event\Observer $observer): void
+    {
+        $customerId = $observer->getEvent()->getData('customer_id');
+        //Call to some query api
+        $messagePayload = $this->customerQueryApi->get($customerId);
+        $message = $this->webHookMessageFactory->create('customer_register', $messagePayload);
+        $this->webHookDispatcher->send($message);
+        
+    }   
+}
+```
+:warning: It's very important to use query API for data retreval purposes where applicable. Using query API allows to apply service business logic on top of changed data, so we can reduce copy-paste of business logic in consumers. Catalog example: `visibility` field + category permissions functionality + child product functionality affect the actual product visibility on store-front, so we need to send `category_assigment` event instead of providing the raw `visibility` field that should be processed on consumer's side in exactly same manner as Magento processes.:warning:
+
+Webhook messages are created using provided payload that must implement `JsonSerializable` interface. Example:
+```php
+/**
+ * The output of jsonSerialize is kind of api as external systems will rely on it's content
+ * @api
+ */
+class CustomerUpdatePayload implements \JsonSerializable 
```

Comment:
in webhooks.xml you mention Magento\Customer\Api\Data\CustomerQueryInterface which looks like a regular DTO but here you say that payloads must return array of scalars to be serialized to JSON. What if we decide to use multiple gateway to send these messages and some of them don't employ JSON? I think it will be better to have DTOs as payloads and different gateway services serializing them as they see fit

Reply:
DTO is not really important, I planned to utilize JsonSerializable contract instead. JsonSerialize doesn't produce json in output, most probably it's just a php array. Different gateways can use different formats for communications.